Add Options support to JarInJar locator

LexManos opened the pull request:
This is a re-write of the JarInJar locator to both be more efficient by leveraging roimfs.
As well as adding the ability for us to ship 'options' for JarInJar dependencies.
Yall know the history of this, MixinExtras and people being to vain/lazy to JiJ it themselves. 
See my big wall of a comment in JarInJarDependencyLocator for some explinations of how it works and why I decided to do things the way I did.

Requires roimfs, which i'll publish after adding benchmarks.
Doesn't strictly need https://github.com/MinecraftForge/SecureModules/commit/c5ead60f32198ff669167d61f74a0933689df8cf but that fixes some issues/makes it not wrap in UnionFileSytem.
Requires https://github.com/MinecraftForge/JarJar/pull/11 which will be pulled once this is all finalized.

Compatibility wise there shouldn't be any concerns. As all code changes are private inside the locator, and it supports the old JIJ metadata file format. 
There are behavior changes tho, namely the use of roimfs instead of JarIJar's FileSystem. 
But the FileSystem is still shipped, so people can still use it. And Mod were designed to be on any FileSystem. So it shouldn't be an issue. Tho in a future breaking window, I suggest we stop shipping JIJFS and let that project die.

Thanks @Jonathing for the gradle work to build the json file.

Jonathing commented on build_shared.gradle:
```diff
@@ -12,7 +12,7 @@ repositories {
     maven gradleutils.forgeMaven
     maven gradleutils.minecraftLibsMaven
-    //mavenLocal()
+    mavenLocal()
```
Reminder to comment this out once the PR is finalized.

Jonathing commented on settings.gradle:
```diff
@@ -36,7 +36,10 @@ dependencyResolutionManagement {
             library('eventbus-validator', 'net.minecraftforge:eventbus-validator:7.0-beta.12')
             library('mergetool-api', 'net.minecraftforge:mergetool-api:1.0')
 
+            library('roimfs', 'net.minecraftforge:roimfs:0.0.0') // ReadOnlyInMemoryFileSystem - Used for JarInJar extraction at runtime
```
Reminder to publish roimfs and update its version number here.

Jonathing commented on settings.gradle:
```diff
@@ -76,7 +79,7 @@ dependencyResolutionManagement {
 
             // Jar in Jar FileSystem
-            version('jarjar', '0.3.26')
+            version('jarjar', '0.3.39')
```
Reminder to check the version number here once the related PR has been merged.

Jonathing commented on fmlloader/build.gradle:
```diff
@@ -103,11 +104,36 @@ tasks.register('writeForgeVersionJson') {
+tasks.register('jarJarOptionsJson', net.minecraftforge.forge.tasks.JarJarMetadataOptions) {
+    metadataFile = project.file('src/main/resources/jarjar_options.json')
+    // This is resolved too early by cpw.mods.modlauncher.TransformationServicesHandler.discoverServices(DiscoveryData) But eventually...
+    //add(libs.mixin, 'org/spongepowered/asm/mixin/Mixin.class') 
```
A man can dream 😭

LexManos replied:
I've looked into this, the issue is worse then i thought. Mixin is hooked into cpw.mods.modlauncher.serviceapi.ILaunchPluginService which means it starts bootstrapping before any mods directory discovery is done. 

I have some potential ideas on how to fix, but would involve going all the way back to bootstrap, and delaying the ILaunchPluginService system that we also use.

I'm officially making it as out of scope of this PR and something to solve for the future.

Jonathing replied:
Not worth looking into Mixin until new loader. It is unfortunately rather limiting what we can do with it at the current moment, and I know Paint's had some growing pains with it when designing the new loader. It'll need its own week with a looooot of alcohol to go with it.

Jonathing commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
```diff
@@ -5,160 +5,253 @@
-               .toList();
+        if (selected.isEmpty()) {
+            LOGGER.info("No dependencies to load found. Skipping!");
+            return Collections.emptyList();
```
inb4 Paint, but I think `List.of()` might have a JIT optimization. Worth double-checking.

PaintNinja replied:
Doesn't matter in this specific case. Main use for `List.of()` over `Collections.emptyList()` is when you have a Stream#toList(), List.of(...) and/or List#copyOf so that it stays as two types that can be returned by the method, or when you want null hostility.

But this method always returns either an `ArrayList` or `Collections.emptyList()`, so perf-wise it's fine and API-wise having it consistently not throw when doing list.contains(null) is good.

Jonathing commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
Ideally, for new loader, we stick with one path and don't need to worry about edge cases and weird backwards compatibility like we do for current loader. So, it would likely be extracting jars.

If for some black magic reason roimfs actually ends up being faster than extracting and hashing nested jars, that could also be something else we look into. But I doubt it.

LexManos replied:
I would vote for extracting files as well because memory wise its more efficient. As ZipFileSystem can use a mapped FileChannel which means it doesn't need to load the entire thing to memory. TECHNICALLY its slower because it means having to read from the disc when reading entries and disc IO is slower then memory access. But it makes life A LOT more simple compatibility wise.

And yes, removing the ability to nest jars more then one deep, and adding some extra metadata to the json will make things A LOT faster. It would allow us to only extract the files that we are going to use instead of having to extract them to see if they contain more deps to load. I could go on a rant about the pros and cons of each choice. But thats for a future time.

PaintNinja commented on buildSrc/src/main/groovy/net/minecraftforge/forge/tasks/JarJarMetadataOptions.java:
```diff
@@ -0,0 +1,406 @@
+                throw new IllegalArgumentException("Version is unspecified for: " + dependency.asString, e);
+            }
+        }
+    }
+
+    public interface ResolvedDependencyInfo {
+        void containedJarMetadata(Action<? super ContainedJarMetadataInfo> action);
+
+        void setResource(String resource);
+
+        void setLayer(String layer);
+
+        void setId(String identifier);
+
+        interface ContainedJarMetadataInfo {
+            void setGroup(String group);
+
+            void setName(String name);
+
+            void setPath(String version);
+        }
+    }
```
javadocs?

Jonathing replied:
This is for ForgeDev 6. I do not care about Javadocs, no one is going to be using this. Documentation will be accounted for when the JarJar plugin is updated to account for the options editions, or if ForgeDev 7 will include them.

LexManos replied:
Ya this just a quick and dirty 'make it work' part of buildSrc not end user facing. I'll leave this to Jonathing to resolve in ForgeDev proper.

Jonathing replied:
Hell yeah

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
Missing null check on `mod`, preventing it from logging the error on the next line when mod is null.

LexManos replied:
There should be no code paths that result in that, however to better illustrate that I will move this code inside the try. The layout was a remnant of an old implementation before I wrote roimfs.

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
Minor: these can be merged

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
Double-checked locking without volatile. Add volatile to the `optionsLoaded` boolean or better use AtomicBoolean#compareAndSet/VarHandle#compareAndSet to ensure it only loads once under race/thread contention

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
```diff
@@ -169,5 +262,296 @@ protected String identifyMod(IModFile modFile) {
 
-    private record ModWithVersionRange(IModInfo modInfo, VersionRange versionRange, ArtifactVersion artifactVersion) {}
+    private static final Entry FAILED = new Entry("FAILED ENTRY", Path.of(""), Type.LIBRARY, null);
+    private static class Entry implements Comparable<Entry>{
```
I think this would be cleaner as a private sealed interface and records, but if you don't want to do that, sealed on here and final on the subclasses would be good at least.

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
```diff
@@ -169,5 +262,296 @@ protected String identifyMod(IModFile modFile) {
+
+    private class Selector extends JarSelector<Entry> {
+        private Map<Key, Entry> entries = new HashMap<>();
+        private Map<Entry, Map<String, String>> children = new HashMap<>();
```
```suggestion
    private final class Selector extends JarSelector<Entry> {
        private final HashMap<Key, Entry> entries = new HashMap<>();
        private final HashMap<Entry, HashMap<String, String>> children = new HashMap<>();
```

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
```diff
@@ -169,5 +262,296 @@ protected String identifyMod(IModFile modFile) {
+                byte[] data = Files.readAllBytes(targetPath);
+                String filename = targetPath.getFileName().toString();
+                URI uri = new URI(ROIMFS, "jar-jar-" + nextId() + "/" + filename, null);
+                var roimfs = FileSystems.newFileSystem(uri, Collections.singletonMap("data", data));
+                roimfsPath = roimfs.getPath(filename);
+                var zip = FileSystems.newFileSystem(new URI("jar:" + roimfsPath.toUri()), Collections.emptyMap());
```
Do these filesystems need closing?

LexManos replied:
They are closed, if they end up being not needed, in the scanMods function (under the big comment) We want to keep them option so that we can access them in getResource/getNested. And eventually the resulting SecureJar if the mod is selected.

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
```diff
@@ -169,5 +262,296 @@ protected String identifyMod(IModFile modFile) {
+                    type = IModFile.Type.valueOf(value);
+                else
+                    type = IModFile.Type.MOD;
+            } else if (mf != null) {
```
`mf` is never null because it is set to an empty manifest on line 449 and possibly changed to a populated one on line 453. Either initially set to null and also handle it on line 476, or remove this null check and cache the empty manifest object

PaintNinja commented on fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/JarInJarDependencyLocator.java:
```diff
@@ -169,5 +262,296 @@ protected String identifyMod(IModFile modFile) {
+                id = mf.getMainAttributes().getValue("Automatic-Module-Name");
+
+            if (id == null)
+                id = Paths.get(path).getFileName().toString();
```
Minor:
```suggestion
                id = Path.of(path).getFileName().toString();
```

LexManos commented:
Alright, just to summarize some discussions from discord.
My end is functionally done. So I am now waiting on Jonathing to address the gradle side.
Once that is done this PR will be split into two parts:
First part is just the JarJar changes
Second part is bundeling Mixin Extras.

This will be done to isolate features and make potential backporting easier.

Speaking of backporting. There is no immediate plans to pull this into 1.20.1, However if Llama wanted to attempt it and submit a Pr then he is more then welcome. If not it'd be on the list of things to back port after the new toolchain is done. Along with things like executable jars, and performance fixes. So to be clear, it IS an eventual goal to backport a lot of things to older version when possible.

As for the gradle side. It will **always** be an opt-in system for enabling Mixin or MixinExtras. This is a hard requirement that I am applying. We are trying to get away from the 'fg is magic' because then the side effects can get out of control fairly easily.

However it should be as simple as a one line option in your gradle. The specific is to be determined but something like `fg.enableMixin()` or `apply plugin: 'org.spongepowered.mixin'` or or `implementation fg.mixinExtras` something like that. 

It has also been discussed to 'auto-enable' mixinextras if we detect that a mod has mixins at all. I'll leave the specifics of that magic to Paint. This PR is specifically designed to allow for Forge to package options. Not specifically any option. Hence the split I said in the first paragraph.

Jonathing commented:
The opt-in hard requirement should at least be some sort of declaration of your Mixins config JSON. ForgeGradle Magic can take care of additional behavior from there.

This magic can be refined in future updates, but for now I think just checking for that will suffice.

LexManos commented:
> The opt-in hard requirement should at least be some sort of declaration of your Mixins config JSON. 

That might actually be an option. Tho it would require reading the config as well as the manifest. Again I'll leave this up to Paint for the runtime magic. Either way it'll be simple and we'll have examples.

Jonathing commented:
Please prefix the PR with [1.21.10]. Because I don't have push access, I cannot modify PR details.

We can discuss GitHub permissions at a later date. I've already alluded to the GitHub permissions write-up I've been working on. It's not top priority, but know that I'm aware of it and am working on it.

Jonathing commented:
From @PaintNinja:

> `selector.select()` in JarInJarDependencyLocator#scanMods throws an NPE when ME is added as a dep. To reproduce, delete the `selector.isRequired(entry.meta.identifier())` part of the loop just before it

https://discord.com/channels/1129059589325852724/1129095235889270844/1431671142053183660

LexManos committed:
Add support for Forge shipping with JarInJar options.
These are optional libraries that are only loaded when requested by mods.
Includes new ReadOnlyInMemoryFileSystem to improve performance of JarInJar parsing.

LexManos commented:
The selector NPE has been fixed. It was caused by a state where there were dependencies in the list when they were not asked for.
I have also split this PR into two pieces, one adding the functionality and the other for Mixin Extras.
Once Azure is back this PR will be pulled.
I'll leave the Mixin-Extra specific things to a second PR with Llama/Paint figuring out the specifics.
https://github.com/MinecraftForge/MinecraftForge-Experimental/commit/3c95fa5a0643e6ec5f46a84cc7c46ef210e195b3
Beyond that, there is still need for the gradle magic from Jonathing.

Jonathing commented:
Works for me. I don't know if Azure will be back online before the end of the day, but I'll figure something out with the specifics of what we have.

That being said, the hack I will end up adding for ForgeGradle 6 at first will be pretty volatile. If we don't decide to add a field to Patcher v3 that contains the ME version, I will need to make it best-guess 0.5.0. Additionally, as I've mentioned on the Discord, the magic for FG6 will be very surface-level (`fg.enableMixinExtras()`). I understand the desire from LlamaLad and others for a 0-line application, but I don't think it is possible in FG6's current state.

I'm working to freeze the API for ForgeGradle 7.0 soon, so my attention is going to end up being split between that and this. Lex and Paint are still sitting around waiting to be able to help me with ForgeDev, so once I get the surface-level user-side stuff done, I will need to focus on that so I'm not the only one working on it anymore.
